types/redux-form/v7/lib/FieldArray.d.ts

```diff
@@ -1,29 +1,29 @@
 import { Component, ComponentType } from "react";
 import { Validator } from "../index";
 
-export interface BaseFieldArrayProps<P = {}> {
+export interface BaseFieldArrayProps<P = {}, FieldValue = any> {
     name: string;
-    component: ComponentType<P>;
+    component: ComponentType<WrappedFieldArrayProps<FieldValue> & P>;
     validate?: Validator | Validator[];
     warn?: Validator | Validator[];
     withRef?: boolean;
     props?: P;
     rerenderOnEveryChange?: boolean;
 }
 
-export interface GenericFieldArray<Field, P = {}> extends Component<BaseFieldArrayProps<P> & Partial<P>> {
```

Comment:
@VincentBel, I see @m-b-davis is removing `& Partial<P>` here in favour of using `FieldValue` - would you be able to take a look and see what you think as I don't have the context for why `Partial<P>` was used originally and what potential use-cases might be out there.

Reply:
@huwmartin I think it is necessary here because we can use `FieldArray` like this:

```js
const renderMembers = ({ fields, meta, foo }) => (
  <div>{foo}</div>
)

const FieldArraysForm = ({ foo }) => {
  return <FieldArray name="members" foo={foo} component={renderMembers} />
}
```

Reply:
@VincentBel - I don't believe this is a valid way of using FieldArray. Could you provide an example from the docs where this is supported behaviour? 

From the docs:

![image](https://user-images.githubusercontent.com/15704216/58791587-6253a080-85ea-11e9-9472-fb0358f7dd5d.png)
https://redux-form.com/8.1.0/docs/api/fieldarray.md/

As the props prop is the object which contains extra props to pass to the child, (I think) your example above should be written as :

```
const renderMembers = ({ fields, meta, foo }) => (
  <div>{foo}</div>
)

const FieldArraysForm = ({ foo }) => {
  return <FieldArray name="members" props={{ foo }} component={renderMembers} />
}
```

What are your thoughts on this?

Reply:
Hi, @m-b-davis.

It is just like how we can pass props when using `Field` component ([Field doc](
https://redux-form.com/8.1.0/docs/api/field.md/#-code-props-object-code-optional-)):

![image](https://user-images.githubusercontent.com/6076919/58793258-57b7fc00-8628-11e9-8ccc-100650b7a2e7.png)

I think it is just not documented at the `FieldArray` API page. Here is a demo how we can use it: https://codesandbox.io/s/reduxform-fieldarray-props-bqrgf

Reply:
@VincentBel I see - thanks for the clarification! Didn't realise that both methods were supported here.

I've updated the typings to support both of the use-cases. I've still removed partial - using partial here means that all properties of P are possibly undefined when being consumed by the child.

To keep type-safety the props have to be either of the variety where P is merged with the overall props, or provided as the prop named props. I can't see why you'd ever need to use both at the same time - and providing, for example, foo & props: { foo } would lead to undefined behaviour.

New type def is as follows:

```
export type BaseFieldArrayProps<P = {}, FieldValue = any> = (P | { props: P}) & _BaseFieldArrayProps<P, FieldValue>;

interface _BaseFieldArrayProps<P = {}, FieldValue = any> {
    name: string;
    component: ComponentType<WrappedFieldArrayProps<FieldValue> & P>;
    validate?: Validator | Validator[];
    warn?: Validator | Validator[];
    withRef?: boolean;
    rerenderOnEveryChange?: boolean;
}
```

So after defining a custom field array like so:

```
interface MyFieldValue {
    num: number;
}
interface MyFieldArrayCustomProps {
    foo: string;
    bar: number;
}

const MyCustomArrayField: React.StatelessComponent<MyFieldArrayCustomProps & WrappedFieldArrayProps<MyFieldValue>> = ({
    children,
    fields,
    foo,
    bar
}) => null;

const FieldArrayCustom = FieldArray as new () => GenericFieldArray<MyFieldValue, MyFieldArrayCustomProps>;

```

We can then use it 

```
                            {/* Good */}
                            <FieldArrayCustom
                                name="field10"
                                component={ MyCustomArrayField }
                                props={{
                                    foo: 'bar',
                                    bar: 123
                                }}
                            />

                            {/* Good */}
                            <FieldArrayCustom
                                name="field10"
                                component={ MyCustomArrayField }
                                foo="bar"
                                bar={23}
                            />

                            {/* Property 'bar' is missing in type '{ foo: string; }' but required in type 'MyFieldArrayCustomProps'. */}
                            <FieldArrayCustom
                                name="field10"
                                component={ MyCustomArrayField }
                                props={{
                                    foo: 'bar'
                                }}
                            />
```

Thoughts?

Reply:
Yeah, totally agree.